signer/src/codec.rs

```diff
@@ -1,13 +1,32 @@
 //! # Canonical encoding and decoding for the sBTC signer
 //!
-//! The purpose of this module is to define how to encode and decode
-//! signer messages as byte sequences.
+//! The purpose of this module is to define how to encode and decode signer
+//! messages as byte sequences.
 //!
-//! This is achieved by
+//! ## Codec specification
```

Comment:
nit: while point 2 & 3 are enforced on the read path by prost (IIUC), that is not the case for point 1 (in general, other than for `Signed<SignerMessage>`; eg in nested messages) and maybe also for the last part of point 4.
Should we specify it, to avoid giving the impression that we have such guarantees in place?

Reply:
I thought about mentioning it in these docs (and did here https://github.com/stacks-network/sbtc/pull/1188/commits/8fb948b63ce760f93d2194b452b2ade40be54315 before I removed it). Yeah, I'll mention it in these docs to avoid giving the wrong impression.

Reply:
I think I may have partially misunderstood what you were saying. So I believe `prost` generates code that enforces all 4 points all of the time, but you're right, we do not have such guarantees in place. We do not test against it... maybe we should though... brb.

Reply:
Added tests in https://github.com/stacks-network/sbtc/pull/1188/commits/07f8d0089062e5ac535571b5a665c3776c8af1c0.

Reply:
Sorry, I should have worded it better; having those extra tests is def nice, but I was referring to something slightly different: while it's true (and now tested) that prost enforces those requirements when serializing the messages, we are not enforcing some of those requirements when deserializing:
 - we don't check tag order other than for top level `Signed<SignerMessage>`
 - we don't check that we don't deserialize empty values -- although this may be just in the proto spec so nothing else to do here

While the above shouldn't be important outside the digest/signature computation, I'm worried that as is written it may give the impression that (eg) the tag order check is enforced on deserialization for *any* message.

Reply:
Oh, gotcha. I'll make sure to note that the current deserialization code does not enforce these rules.

Reply:
Updated in https://github.com/stacks-network/sbtc/pull/1188/commits/7224280045181eb0022d5476982667a5351702ce